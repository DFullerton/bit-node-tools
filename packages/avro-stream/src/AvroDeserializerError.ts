import { Message } from 'kafka-node';

export class AvroDeserializerError extends Error {
  constructor(message: string, public chunk: Message, public encoding: string, public originalError?: Error) {
    super(message);
  }
}
