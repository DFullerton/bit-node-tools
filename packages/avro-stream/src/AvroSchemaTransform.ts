import { ForSchemaOptions, Schema, Type } from 'avsc';
import { Transform } from 'stream';
import { SchemaRegistryResolver } from './SchemaRegistryResolver';
import { SchemaResolver } from './types';

export abstract class AvroSchemaTransform extends Transform {
  protected resolver: SchemaResolver;

  constructor(resolver: SchemaResolver | string, private schemaOptions?: Partial<ForSchemaOptions>) {
    super({ objectMode: true });
    this.resolver = typeof resolver === 'string' ? new SchemaRegistryResolver(resolver) : resolver;
  }

  typeForSchema(schema: Schema) {
    return Type.forSchema(schema, { registry: {}, ...this.schemaOptions });
  }
}
