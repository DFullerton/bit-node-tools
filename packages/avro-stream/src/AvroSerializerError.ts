import { AvroProduceRequest } from './types';

export class AvroSerializerError extends Error {
  constructor(
    message: string,
    public chunk: AvroProduceRequest,
    public encoding: string,
    public originalError?: Error,
  ) {
    super(message);
  }
}
