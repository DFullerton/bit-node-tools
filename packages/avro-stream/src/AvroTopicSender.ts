import { Readable, ReadableOptions } from 'stream';
import { AvroProduceRequest, AvroTopicSenderOptions } from './types';

export class AvroTopicSender<TMessage = any> extends Readable {
  readonly options: AvroTopicSenderOptions;

  constructor(options: AvroTopicSenderOptions & ReadableOptions) {
    const { schema, topic, partition, key, ...readableOptions } = options;
    super({ ...readableOptions, objectMode: true });
    this.options = { schema, topic, partition, key };
  }

  send(...messages: TMessage[]) {
    const produceRequest: AvroProduceRequest<TMessage> = { ...this.options, messages };
    this.push(produceRequest);
    return this;
  }

  close() {
    this.push(null);
    return this;
  }
}
