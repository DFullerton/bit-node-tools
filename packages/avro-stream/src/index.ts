export { AvroDeserializer } from './AvroDeserializer';
export { AvroSerializer } from './AvroSerializer';
export { SchemaRegistryResolver } from './SchemaRegistryResolver';
export { SchemaResolver } from './types';
export { constructMessage, deconstructMessage } from './message';
