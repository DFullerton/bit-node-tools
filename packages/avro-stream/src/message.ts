export interface AvroMessage {
  schemaId: number;
  buffer: Buffer;
}

export const deconstructMessage = (buffer: Buffer): AvroMessage => {
  return { schemaId: buffer.readInt32BE(1), buffer: buffer.slice(5) };
};

export const constructMessage = ({ schemaId, buffer }: AvroMessage): Buffer => {
  const prefix = Buffer.alloc(5);
  prefix.writeUInt8(0, 0);
  prefix.writeUInt32BE(schemaId, 1);

  return Buffer.concat([prefix, buffer]);
};
