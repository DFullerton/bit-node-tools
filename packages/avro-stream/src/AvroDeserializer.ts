import { Type } from 'avsc';
import { Message } from 'kafka-node';
import { Transform, TransformCallback } from 'stream';
import { deconstructMessage } from './message';
import { SchemaRegistryResolver } from './SchemaRegistryResolver';
import { SchemaResolver } from './types';

export class AvroDeserializer extends Transform {
  private resolver: SchemaResolver;

  constructor(resolver: SchemaResolver | string) {
    super({ objectMode: true });
    this.resolver = typeof resolver === 'string' ? new SchemaRegistryResolver(resolver) : resolver;
  }

  async _transform(message: Message, encoding: string, callback: TransformCallback) {
    try {
      if (typeof message.value === 'string') {
        throw new Error(`ConsumerGroupStream for topic "${message.topic}" must set the encoding to "buffer"`);
      }

      const { schemaId, buffer } = deconstructMessage(message.value);
      const schema = await this.resolver.fromId(schemaId);
      callback(undefined, Type.forSchema(schema).fromBuffer(buffer));
    } catch (error) {
      callback(error);
    }
  }
}
