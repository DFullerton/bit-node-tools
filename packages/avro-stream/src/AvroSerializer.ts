import { Type } from 'avsc';
import { Transform, TransformCallback } from 'stream';
import { constructMessage } from './message';
import { SchemaRegistryResolver } from './SchemaRegistryResolver';
import { AvroProduceRequest, SchemaResolver } from './types';

export class AvroSerializer extends Transform {
  private resolver: SchemaResolver;

  constructor(resolver: SchemaResolver | string) {
    super({ objectMode: true });
    this.resolver = typeof resolver === 'string' ? new SchemaRegistryResolver(resolver) : resolver;
  }

  async _transform(request: AvroProduceRequest, encoding: string, callback: TransformCallback) {
    try {
      const type = Type.forSchema(request.schema);
      const schemaId = await this.resolver.toId(request.topic, request.schema);

      callback(undefined, {
        ...request,
        messages: request.messages.map(message => constructMessage({ schemaId, buffer: type.toBuffer(message) })),
      });
    } catch (error) {
      callback(error);
    }
  }
}
