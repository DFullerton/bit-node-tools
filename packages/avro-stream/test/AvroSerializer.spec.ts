import { ReadableMock, WritableMock } from 'stream-mock';
import { AvroSerializer, SchemaResolver } from '../src';

describe('Integration test', () => {
  it('Test Serialier', async () => {
    const sourceData = [
      {
        topic: 'test-topic-1',
        schema: {
          type: 'record',
          name: 'TestSchema1',
          fields: [{ name: 'accountId', type: 'string' }],
        },
        messages: [{ accountId: '111' }, { accountId: '222' }],
      },
      {
        topic: 'test-topic-2',
        schema: {
          type: 'record',
          name: 'TestSchema2',
          fields: [{ name: 'effectiveEnrollmentDate', type: { type: 'int', logicalType: 'date' } }],
        },
        messages: [{ effectiveEnrollmentDate: 17687 }, { effectiveEnrollmentDate: 17567 }],
      },
    ];

    const schemaResolverMock: SchemaResolver = {
      toId: jest
        .fn()
        .mockResolvedValueOnce(1)
        .mockResolvedValueOnce(2),
      fromId: jest.fn(),
    };

    const sourceStream = new ReadableMock(sourceData, { objectMode: true });
    const sinkStream = new WritableMock({ objectMode: true });
    const serializer = new AvroSerializer(schemaResolverMock);

    sourceStream.pipe(serializer).pipe(sinkStream);

    await new Promise(resolve => {
      sinkStream.on('finish', () => {
        expect(schemaResolverMock.toId).toHaveBeenCalledTimes(2);
        expect(schemaResolverMock.toId).toHaveBeenCalledWith(sourceData[0].topic, sourceData[0].schema);
        expect(schemaResolverMock.toId).toHaveBeenCalledWith(sourceData[1].topic, sourceData[1].schema);
        expect(sinkStream.data).toMatchSnapshot();
        resolve();
      });
    });
  });
});
