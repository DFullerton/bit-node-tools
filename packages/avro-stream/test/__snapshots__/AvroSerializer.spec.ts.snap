// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test Test Serialier 1`] = `
Array [
  Object {
    "messages": Array [
      Object {
        "data": Array [
          0,
          0,
          0,
          0,
          1,
          6,
          49,
          49,
          49,
        ],
        "type": "Buffer",
      },
      Object {
        "data": Array [
          0,
          0,
          0,
          0,
          1,
          6,
          50,
          50,
          50,
        ],
        "type": "Buffer",
      },
    ],
    "schema": Object {
      "fields": Array [
        Object {
          "name": "accountId",
          "type": "string",
        },
      ],
      "name": "TestSchema1",
      "type": "record",
    },
    "topic": "test-topic-1",
  },
  Object {
    "messages": Array [
      Object {
        "data": Array [
          0,
          0,
          0,
          0,
          2,
          174,
          148,
          2,
        ],
        "type": "Buffer",
      },
      Object {
        "data": Array [
          0,
          0,
          0,
          0,
          2,
          190,
          146,
          2,
        ],
        "type": "Buffer",
      },
    ],
    "schema": Object {
      "fields": Array [
        Object {
          "name": "effectiveEnrollmentDate",
          "type": Object {
            "logicalType": "date",
            "type": "int",
          },
        },
      ],
      "name": "TestSchema2",
      "type": "record",
    },
    "topic": "test-topic-2",
  },
]
`;
